### PR TITLE
feat: add experimental optimistic provide

### PR DESCRIPTION
**Pull request by guseggert**
This adds the ability to enable "optimistic provide" to the default
DHT client, which enables faster provides and reprovides.

For more information about optimistic provide, see:

https://protocollabs.notion.site/Optimistic-Provide-2c79745820fa45649d48de038516b814

Note that this feature only works when using non-custom router
types. This does not include the ability to enable optimistic provide
on custom routers for now, to minimize the footprint of this
experimental feature. We intend on continuing to test this and improve
the UX, which may or may not involve adding configuration for it to
custom routers. We also plan on refactoring/redesigning custom routers
more broadly so I don't want this to add more effort for maintainers
and confusion for users.

### CODE REVIEW

**Comment by guseggert** [comment]
This needs docs in `docs/experimental-features.md` and a changelog entry

File: config/experiments.go [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+	FilestoreEnabled              bool
+	UrlstoreEnabled               bool
+	ShardingEnabled               bool `json:",omitempty"` // deprecated by autosharding: https://github.com/ipfs/kubo/pull/8527
+	GraphsyncEnabled              bool
+	Libp2pStreamMounting          bool
+	P2pHttpProxy                  bool //nolint
+	StrategicProviding            bool
+	AcceleratedDHTClient          bool
+	OptimisticProvide             bool
+	OptimisticProvideJobsPoolSize int
```
Comment (aschmahmann) [inline config/experiments.go:13]: Two questions:
1. Can you also add explanations of these into the docs folder?
2. Are these supposed to be global or per-DHT router? (or maybe they're supposed to be per-dht but that's pending the fixes to custom routing)

Comment (guseggert) [reply config/experiments.go:13]: 1) Yes @dennis-tra is going to fill out the docs for this

2) I think all we _really_ care about is the WAN DHT here right? I've tried to just enable opt prov on everything when this is enabled, _unless_ you are using a `custom` router.

Comment (dennis-tra) [reply config/experiments.go:13]: Added the description 👍

**Review by BigLep — APPROVED** [review]
Aprovig because:
1. Logically the changes made sense to me.
2. This is an experimental feature
3. The docs looked fine
4. I know others have looked at this too

**Comment by BigLep** [comment]
Two other thoughts:
1. I assume you'll add to the changelog in https://github.com/ipfs/kubo/pull/9799
2. Are there other issues or PR's we should link to (e.g., https://github.com/libp2p/go-libp2p-kad-dht/pull/783 )
3. Also, in the experimental feature docs maybe get links to more info like https://protocollabs.notion.site/Optimistic-Provide-2c79745820fa45649d48de038516b814 and https://github.com/libp2p/notes/issues/25

**Pull request merged**